scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryService.scala

```diff
@@ -1,9 +1,798 @@
 package io.iohk.scalanet.discovery.ethereum.v4
 
-/** Implement the stateful discovery logic:
-  * - maintain the state of K-buckets
-  * - return node candidates for the rest of the system
-  * - bond with the other nodes
-  * - respond to incoming requests
+import cats.effect.{Clock, Resource}
+import cats.effect.concurrent.{Deferred, Ref}
+import cats.implicits._
+import io.iohk.scalanet.discovery.crypto.{PrivateKey, SigAlg}
+import io.iohk.scalanet.discovery.ethereum.{Node, EthereumNodeRecord}
+import io.iohk.scalanet.discovery.hash.Hash
+import io.iohk.scalanet.kademlia.KBuckets
+import java.net.InetAddress
+import monix.eval.Task
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+import scodec.{Codec, Attempt}
+import com.typesafe.scalalogging.LazyLogging
+import scala.collection.immutable.SortedSet
+import io.iohk.scalanet.kademlia.XorOrdering
+
+/** Represent the minimal set of operations the rest of the system
+  * can expect from the service to be able to talk to other peers.
   */
-trait DiscoveryService[A] {}
+trait DiscoveryService {
+
+  /** Try to look up a node either in the local cache or
+    * by performing a recursive lookup on the network. */
+  def getNode(nodeId: Node.Id): Task[Option[Node]]
+
+  /** Return all currently bonded nodes. */
+  def getNodes: Task[Set[Node]]
+
+  /** Try to get the ENR record of the given node to add it to the cache. */
+  def addNode(node: Node): Task[Unit]
+
+  /** Remove a node from the local cache. */
+  def removeNode(nodeId: Node.Id): Task[Unit]
+
+  /** Update the local node with an updated external address,
+    * incrementing the local ENR sequence.
+    */
+  def updateExternalAddress(ip: InetAddress): Task[Unit]
+
+  /** The local node representation. */
+  def getLocalNode: Task[Node]
+}
+
+object DiscoveryService {
+  import DiscoveryRPC.{Call, Proc}
+  import DiscoveryNetwork.Peer
+
+  type ENRSeq = Long
+  type Timestamp = Long
+  type StateRef[A] = Ref[Task, State[A]]
+
+  /** Implement the Discovery v4 protocol:
+    *
+    * https://github.com/ethereum/devp2p/blob/master/discv4.md
+    *
+    * - maintain the state of K-buckets
+    * - return node candidates for the rest of the system
+    * - bond with the other nodes
+    * - respond to incoming requests
+    * - periodically try to discover new nodes
+    * - periodically ping nodes
+    */
+  def apply[A](
+      privateKey: PrivateKey,
```

Comment:
As i saw it is only used in `updateExternalAddress` maybe we can  pass it there as as function param i.e have 
` override def updateExternalAddress(ip: InetAddress, privateKey: PrivateKey)`. That way user of library would not be forced to keep private key in memory during whole time.

Reply:
Well the `DiscoveryNetwork` also needs a copy of the private key since it has to sign every outgoing packet. Maybe you'd be happier with a `privateKeyReader: Task[PrivateKey]` that may or may not load the key, on demand? 

The reason I hesitate to add it to the `updateExternalAddress` is that it's a method that looks like would be called by a pure networking component that somehow learns about our external IP: why should it know the key, or the fact that this component needs to sign an ENR?

Reply:
but even the docs of this method says `incrementing the local ENR sequence.` which kinda implies sigining enr sequence with new number. 

To be honest from mantis perspective it is not that important for now, as we are keeping private key in memory there eitherway.